File: frontend/src/types/incidents.ts
export const INCIDENT_STATUSES = [
  "NEW",
  "TRIAGE",
  "INVESTIGATION",
  "CONTAINMENT",
  "ERADICATION",
  "RECOVERY",
  "CLOSED",
] as const;

export const INCIDENT_TRANSITIONS: Record<string, string[]> = {
  NEW: ["TRIAGE", "INVESTIGATION", "CLOSED"],
  TRIAGE: ["INVESTIGATION", "CONTAINMENT", "CLOSED", "NEW"],
  INVESTIGATION: ["CONTAINMENT", "ERADICATION", "TRIAGE", "CLOSED"],
  CONTAINMENT: ["ERADICATION", "INVESTIGATION", "CLOSED"],
  ERADICATION: ["RECOVERY", "CONTAINMENT", "INVESTIGATION", "CLOSED"],
  RECOVERY: ["CLOSED", "ERADICATION", "INVESTIGATION"],
  CLOSED: ["INVESTIGATION", "TRIAGE"],
};

export const TASK_STATUSES = ["open", "in_progress", "done"] as const;
export const INCIDENT_SEVERITIES = ["low", "medium", "high", "critical"] as const;
export const INCIDENT_PRIORITIES = ["P1", "P2", "P3", "P4"] as const;

export interface IncidentPublic {
  id: string;
  display_id: string;
  title: string;
  description: string | null;
  severity: string;
  priority: string;
  status: string;
  analyst_id: string | null;
  resolution: string | null;
  lessons_learned: string | null;
  detected_at: string | null;
  contained_at: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface LinkedAlert {
  id: string;
  display_id: string;
  title: string;
  severity: string;
  risk_score: number;
  status: string;
}

export interface LinkedEntities {
  assets: Record<string, unknown>[];
  iocs: Record<string, unknown>[];
  users: Record<string, unknown>[];
}

export interface IncidentDetail extends IncidentPublic {
  alerts: LinkedAlert[];
  linked: LinkedEntities;
}

export interface IncidentCounts {
  by_status: Record<string, number>;
  open_total: number;
}

export interface IncidentNote {
  id: string;
  author_id: string | null;
  body: string;
  created_at: string;
}

export interface IncidentTask {
  id: string;
  title: string;
  description: string | null;
  assignee_id: string | null;
  status: string;
  due_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface TimelineEntry {
  kind: string;
  summary: string;
  detail: Record<string, unknown>;
  actor_id: string | null;
  occurred_at: string;
}

export interface IncidentEvidenceDocument {
  event_id: string;
  found: boolean;
  document: Record<string, unknown> | null;
}

export interface IncidentEvidenceResponse {
  incident_id: string;
  display_id: string;
  total_event_ids: number;
  resolved: number;
  missing_event_ids: string[];
  documents: IncidentEvidenceDocument[];
}

File: backend/app/schemas/incidents.py
"""Incident API schemas (spec §22 `/incidents`, spec §14's workflow)."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.incidents import INCIDENT_PRIORITIES, INCIDENT_SEVERITIES


class IncidentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=300)
    description: str | None = Field(default=None, max_length=10_000)
    severity: str = "medium"
    priority: str | None = None
    # Alerts to promote into the case at creation time. Promoting later is
    # POST /incidents/{id}/alerts, so this is a convenience, not the only path.
    alert_ids: list[uuid.UUID] = Field(default_factory=list, max_length=50)

    @field_validator("severity")
    @classmethod
    def _known_severity(cls, value: str) -> str:
        if value not in INCIDENT_SEVERITIES:
            raise ValueError(f"severity must be one of {', '.join(INCIDENT_SEVERITIES)}")
        return value

    @field_validator("priority")
    @classmethod
    def _known_priority(cls, value: str | None) -> str | None:
        if value is not None and value not in INCIDENT_PRIORITIES:
            raise ValueError(f"priority must be one of {', '.join(INCIDENT_PRIORITIES)}")
        return value


class IncidentUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=300)
    description: str | None = Field(default=None, max_length=10_000)
    severity: str | None = None
    priority: str | None = None
    analyst_id: uuid.UUID | None = None
    lessons_learned: str | None = Field(default=None, max_length=20_000)

    @field_validator("severity")
    @classmethod
    def _known_severity(cls, value: str | None) -> str | None:
        if value is not None and value not in INCIDENT_SEVERITIES:
            raise ValueError(f"severity must be one of {', '.join(INCIDENT_SEVERITIES)}")
        return value

    @field_validator("priority")
    @classmethod
    def _known_priority(cls, value: str | None) -> str | None:
        if value is not None and value not in INCIDENT_PRIORITIES:
            raise ValueError(f"priority must be one of {', '.join(INCIDENT_PRIORITIES)}")
        return value


class TransitionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: str
    note: str | None = Field(default=None, max_length=2000)


class LinkAlertRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    alert_id: uuid.UUID


class LinkAssetRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    asset_id: uuid.UUID


class LinkIocRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ioc_id: uuid.UUID


class LinkUserRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    username: str = Field(min_length=1, max_length=300)


class NoteRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    body: str = Field(min_length=1, max_length=5000)


class NotePublic(BaseModel):
    id: str
    author_id: str | None
    body: str
    created_at: datetime


class TaskCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=300)
    description: str | None = Field(default=None, max_length=5000)
    assignee_id: uuid.UUID | None = None
    due_at: datetime | None = None


class TaskUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=300)
    description: str | None = None
    assignee_id: uuid.UUID | None = None
    status: str | None = None
    due_at: datetime | None = None

    @field_validator("status")
    @classmethod
    def _known_status(cls, value: str | None) -> str | None:
        from app.models.incidents import TASK_STATUSES

        if value is not None and value not in TASK_STATUSES:
            raise ValueError(f"status must be one of {', '.join(TASK_STATUSES)}")
        return value


class TaskPublic(BaseModel):
    id: str
    title: str
    description: str | None
    assignee_id: str | None
    status: str
    due_at: datetime | None
    completed_at: datetime | None
    created_at: datetime


class TimelineEntryPublic(BaseModel):
    kind: str
    summary: str
    detail: dict[str, Any]
    actor_id: str | None
    occurred_at: datetime


class LinkedAlertPublic(BaseModel):
    id: str
    display_id: str
    title: str
    severity: str
    risk_score: int
    status: str


class LinkedEntities(BaseModel):
    assets: list[dict[str, Any]]
    iocs: list[dict[str, Any]]
    users: list[dict[str, Any]]


class IncidentPublic(BaseModel):
    id: str
    display_id: str
    title: str
    description: str | None
    severity: str
    priority: str
    status: str
    analyst_id: str | None
    resolution: str | None
    lessons_learned: str | None
    detected_at: datetime | None
    contained_at: datetime | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class IncidentDetail(IncidentPublic):
    alerts: list[LinkedAlertPublic]
    linked: LinkedEntities


class IncidentCounts(BaseModel):
    by_status: dict[str, int]
    open_total: int


class EvidenceDocument(BaseModel):
    event_id: str
    found: bool
    document: dict[str, Any] | None = None


class EvidenceResponse(BaseModel):
    incident_id: str
    display_id: str
    total_event_ids: int
    resolved: int
    missing_event_ids: list[str]
    documents: list[EvidenceDocument]

